src/Http/Response.php

```diff
@@ -3,8 +3,396 @@
+   *
+   * @return bool
+   */
+  protected function isEmpty() {
+    return in_array($this->statusCode, array(204, 304));
+  }
+
+  /**
+   * Is response informative?
+   *
+   * @return bool
+   */
+  protected function isInformational() {
+    return $this->statusCode >= 100 && $this->statusCode < 200;
+  }
+
+  /**
+   * Is response successful?
+   *
+   * @return bool
+   */
+  protected function isSuccessful() {
+    return $this->statusCode >= 200 && $this->statusCode < 300;
+  }
+
+  /**
+   * Is response invalid?
+   *
+   * @return bool
+   */
+  protected function isInvalid() {
+    return $this->statusCode < 100 || $this->statusCode >= 600;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function prepare(RequestInterface $request) {
+    $headers = $this->headers;
+    if ($this->isInformational() || $this->isEmpty()) {
+      $this->setContent(NULL);
+      $headers->remove('Content-Type');
+      $headers->remove('Content-Length');
+    }
+    else {
+      // Content-type based on the Request. The content type should have been
+      // set in the RestfulFormatter.
+
+      // Fix Content-Type
+      $charset = $this->charset ?: 'UTF-8';
+      $content_type = $headers->get('Content-Type')->getValueString();
+      if (stripos($content_type, 'text/') === 0 && stripos($content_type, 'charset') === FALSE) {
+        // add the charset
+        $headers->add(HttpHeader::create('Content-Type', $content_type . '; charset=' . $charset));
+      }
+      // Fix Content-Length
+      if ($headers->has('Transfer-Encoding')) {
+        $headers->remove('Content-Length');
+      }
+      if ($request->getMethod() == Request::METHOD_HEAD) {
+        // cf. RFC2616 14.13
+        $length = $headers->get('Content-Length')->getValueString();
+        $this->setContent(NULL);
+        if ($length) {
+          $headers->add(HttpHeader::create('Content-Length', $length));
+        }
+      }
+    }
+    // Fix protocol
+    $server_info = $request->getServer();
+    if ($server_info['SERVER_PROTOCOL'] != 'HTTP/1.0') {
+      $this->setProtocolVersion('1.1');
+    }
+    // Check if we need to send extra expire info headers
+    if ($this->getProtocolVersion() == '1.0' && $this->headers->get('Cache-Control')->getValueString() == 'no-cache') {
+      $this->headers->add(HttpHeader::create('pragma', 'no-cache'));
+      $this->headers->add(HttpHeader::create('expires', -1));
+    }
+    $this->ensureIEOverSSLCompatibility($request);
+  }
+
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setContent($content) {
+    if ($content !== NULL && !is_string($content) && !is_numeric($content) && !is_callable(array($content, '__toString'))) {
+      throw new InternalServerErrorException(sprintf('The Response content must be a string or object implementing __toString(), "%s" given.', gettype($content)));
+    }
+    $this->content = (string) $content;
+    $this->headers->add(HttpHeader::create('Content-Length', strlen($this->content)));
```

Comment:
We're just copying some of methods, I don't think that justifies importing the whole http foundation component. I copied the gotchas ☺